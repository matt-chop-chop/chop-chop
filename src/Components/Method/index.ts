export { default as Method } from "./Method";
