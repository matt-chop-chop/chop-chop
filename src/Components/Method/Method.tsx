import { Box, Heading, ToastProps, useToast } from "@chakra-ui/react";
import { useEffect } from "react";
import { MethodStep, Toast } from "@/Components";
import React from "react";

export type MethodProps = {
  instructions: string[];
};

const Method = ({ instructions }: MethodProps) => {
  const [checkboxes, setCheckboxes] = React.useState<boolean[]>(
    new Array(instructions.length).fill(false)
  );
  const toast = useToast();

  useEffect(() => {
    if (checkboxes.every((checkbox) => checkbox === true)) {
      toast({
        status: "success",
        isClosable: true,
        position: "top-right",
        render: (props: ToastProps) => (
          <Toast
            onClose={props.onClose}
            text="Congratulations, you completed the recipe! Bon AppÃ©tit!"
          />
        ),
      });
    }
  }, [checkboxes, toast]);

  const updateCheckboxes = (checked: boolean, index: number) => {
    const updatedCheckboxes = checkboxes.map((checkbox, i) => {
      if (i === index) {
        return checked;
      } else {
        return checkbox;
      }
    });

    setCheckboxes(updatedCheckboxes);
  };

  if (!instructions.length) return null;

  return (
    <Box my={5}>
      <Heading as="h2" mb={4} variant="heading-medium">
        Method
      </Heading>
      {instructions.map((instruction, index) => (
        <MethodStep
          key={index}
          step={instruction}
          checked={checkboxes[index]}
          setChecked={(checked) => updateCheckboxes(checked, index)}
        />
      ))}
    </Box>
  );
};

export default Method;
