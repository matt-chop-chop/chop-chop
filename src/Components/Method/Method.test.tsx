import { shallow } from "enzyme";
import React from "react";
import Method, { MethodProps } from "./Method";

const props: MethodProps = {
  instructions: ["Test instruction 1", "Test instruction 2"],
};

const useState = jest.spyOn(React, "useState");

describe("Method", () => {
  it("is exported from Method", () => {
    expect(Method).toBeDefined();
  });

  it("renders as expected ", () => {
    const wrapper = shallow(<Method {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("calls setState when a MethodStep setChecked is triggered", () => {
    useState.mockClear();

    const setCheckboxes = jest.fn();
    useState.mockReturnValue([[false, false], setCheckboxes]);

    const wrapper = shallow(<Method {...props} />);

    const MethodStep = wrapper.find("MethodStep");
    expect(MethodStep).toHaveLength(2);

    const setChecked = MethodStep.first().prop("setChecked") as (
      checkboxes: boolean
    ) => void;
    setChecked(true);
    expect(setCheckboxes).toHaveBeenCalledWith([true, false]);
  });

  it("returns null if the list of instructions is empty ", () => {
    const wrapper = shallow(<Method instructions={[]} />);

    expect(wrapper.type()).toBeNull();
  });
});
