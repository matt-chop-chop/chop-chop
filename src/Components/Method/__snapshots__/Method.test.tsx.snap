// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Method renders as expected  1`] = `
<Box
  my={5}
>
  <Heading
    as="h2"
    mb={4}
    variant="heading-medium"
  >
    Method
  </Heading>
  <MethodStep
    checked={false}
    key="0"
    setChecked={[Function]}
    step="Test instruction 1"
  />
  <MethodStep
    checked={false}
    key="1"
    setChecked={[Function]}
    step="Test instruction 2"
  />
</Box>
`;
